File: 1 - metodosBusca/pacoteBusca_JomiHubner/src/busca/BuscaRecursiva.java
package busca;



/**
 *   Algoritmos de Busca (geral, qquer problema)
 *
 *   Busca a solução por busca em profundidade.
 *   Implementação recursiva.
 *
 *   @author Jomi Fred Hübner
 */
public class BuscaRecursiva extends BuscaProfundidade {
     
    /** busca sem mostrar status */
    public BuscaRecursiva() {
    }
    
    /** busca mostrando status */
    public BuscaRecursiva(MostraStatusConsole ms) {
        super(ms);
    }
    
    /**
     * Busca a solução por busca em profundidade.
     *                              ------------
     *
     * Como o algoritmo é recursivo, não precisa da lista de abertos.
     *
     * max é a profundidade máxima de busca.
     */
    public Nodo busca(Estado inicial) {
        status.inicia();
        initFechados();
        Nodo n = buscaR(new Nodo(inicial, null));
        status.termina(n != null);
        return n;
    }
    
    public Nodo buscaR(Nodo corrente) {
        if (corrente == null) {
            return null;
        }
        status.explorando(corrente, 0);
        if (corrente.estado.ehMeta()) {
            return corrente;
        }
        if (corrente.getProfundidade() > profMax || parar) {
            return null;
        }
        for (Nodo s: sucessores(corrente)) {
            Nodo n = buscaR(s);
            if (n != null) {
                return n;
            }
        }
        return null;
    }   
    
    
    public String toString() {
    	return "BP - Busca em Profundidade (m="+profMax+") - versão recursiva";
    }
}

File: 1 - metodosBusca/python/app/core/busca/busca_recursiva.py
from .busca_profundidade import BuscaProfundidade
from .nodo import Nodo
class BuscaRecursiva(BuscaProfundidade):
    def __init__(self, mostra_status_console=None):
        super().__init__(mostra_status_console)

    def busca(self, inicial):
        self.status.inicia()
        self.init_fechados()
        n = self.busca_r(Nodo(inicial, None))
        self.status.termina(n is not None)
        return n
    
    def busca_r(self, corrente):
        if corrente is None:
            return None
        self.status.explorando(corrente, 0)
        if corrente.estado.eh_meta():
            return corrente
        if corrente.get_profundidade() > self.prof_max or self.parar:
            return None
        for s in self.sucessores(corrente):
            n = self.busca_r(s)
            if n is not None:
                return n
        return None
    
    def __str__(self):
        return f"BP - Busca em Profundidade (m={self.prof_max}) - versão recursiva"